fix: edge case for nested generics in a union type

invokermain opened the pull request:
## Description

In `StrawberryUnion.get_type_resolver` prioritise testing against types that are explicitly in the Union type before checking all other known types. This fixes an edge case where StrawberryUnion would pick out an incorrect type that weakly matches (due to nested generics) the resolved object before testing the actual type in the union.

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Resolves #2028.

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

botberry commented:
Thanks for adding the `RELEASE.md` file!

![](https://media.giphy.com/media/xq1FxHkABwW7m/giphy.gif)

Here's a preview of the changelog:

---

Fixed edge case where `Union` types raised an `UnallowedReturnTypeForUnion` error when
returning the correct type from the resolver due to only being able to partially match
nested generic types. This is fixed by prioritising the types explicitly listed in the
Union.

For example the below ([link to playground](https://play.strawberry.rocks/?gist=f7d88898d127e65b12140fdd763f9ef2))
would previously raise the error when querying `two` as `StrawberryUnion` would incorrectly
determine that the resolver returns `Container[TypeOne]`.

```python
import strawberry
from typing import TypeVar, Generic, Union, List, Type

T = TypeVar("T")

@strawberry.type
class Container(Generic[T]):
    items: List[T]

@strawberry.type
class TypeOne:
    attr: str

@strawberry.type
class TypeTwo:
    attr: str

def resolver_one():
    return Container(items=[TypeOne("one")])
    
def resolver_two():
    return Container(items=[TypeTwo("two")])

@strawberry.type
class Query:
    one: Union[Container[TypeOne], TypeOne] = strawberry.field(resolver_one)
    two: Union[Container[TypeTwo], TypeTwo] = strawberry.field(resolver_two)

schema = strawberry.Schema(query=Query)
```


---
Here's the preview release card for twitter:

![](https://i.imgur.com/O2WjzwH.png)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Tim OSullivan for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

botberry commented:
Thanks for adding the `RELEASE.md` file!

![](https://media.giphy.com/media/xq1FxHkABwW7m/giphy.gif)

Here's a preview of the changelog:

---

Fixed edge case where `Union` types raised an `UnallowedReturnTypeForUnion` error when
returning the correct type from the resolver due to only being able to partially match
nested generic types. This is fixed by prioritising the types explicitly listed in the
Union.

For example the below ([link to playground](https://play.strawberry.rocks/?gist=f7d88898d127e65b12140fdd763f9ef2))
would previously raise the error when querying `two` as `StrawberryUnion` would incorrectly
determine that the resolver returns `Container[TypeOne]`.

```python
import strawberry
from typing import TypeVar, Generic, Union, List, Type

T = TypeVar("T")

@strawberry.type
class Container(Generic[T]):
    items: List[T]

@strawberry.type
class TypeOne:
    attr: str

@strawberry.type
class TypeTwo:
    attr: str

def resolver_one():
    return Container(items=[TypeOne("one")])
    
def resolver_two():
    return Container(items=[TypeTwo("two")])

@strawberry.type
class Query:
    one: Union[Container[TypeOne], TypeOne] = strawberry.field(resolver_one)
    two: Union[Container[TypeTwo], TypeTwo] = strawberry.field(resolver_two)

schema = strawberry.Schema(query=Query)
```


---
Here's the preview release card for twitter:

![](https://i.imgur.com/UBs80JM.png)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Tim OSullivan for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

invokermain commented:
I refactored the solution for performance and simplicity.

I benchmarked performance as well, as expected in the best-case scenario (i.e. only the two types used in the Union are present in the schema) it introduces a slight overhead.

However in the worst case scenario (schema defines many types, and all of them appear before the actual resolver type in `type_map.values()`) performance is vastly improved. 

![image](https://user-images.githubusercontent.com/44192816/179752711-f5fdc966-4216-4079-83a7-f47c61887098.png)

This graph tests worse case scenario (although 0 is best case scenario as schema only has the two types):
- The new logic is practically constant with number of types in the schema: `~O(1)`.
- The old logic scales linearly with number of types in the schema: `O(n)`.

patrick91 reviewed (approved):
Nice one! I've updated the release notes slightly and added a TODO comment, it doesn't look like we need to iterate over all the types anymore, right? In any case this can be deal in another PR :)

invokermain commented:
Thanks for merging this, was blocking a ticket at work for me! 

> Nice one! I've updated the release notes slightly and added a TODO comment, it doesn't look like we need to iterate over all the types anymore, right? In any case this can be deal in another PR :)

Yeah I did think that. At first glance you'd think the check:

```python
if return_type not in type_.types:
    raise UnallowedReturnTypeForUnion
```

means we don't need to iterate over the whole `type_map`, i.e. iterating over` type_.types` is enough. But also I don't understand the internals of Strawberry enough to confidently say there isn't an edge case, maybe once you bring generics into the mix. So I did it was just the safest way possible in terms of breaking changes. If I was a more confident person maybe I would change it and just trust the existing unit tests cover all the relevant cases 😄

patrick91 commented:
> Thanks for merging this, was blocking a ticket at work for me!

ohh, next time ping me and I'll try to merge it quicker 😊

 
> > Nice one! I've updated the release notes slightly and added a TODO comment, it doesn't look like we need to iterate over all the types anymore, right? In any case this can be deal in another PR :)
> 
> Yeah I did think that. At first glance you'd think the check:
> 
> ```python
> if return_type not in type_.types:
>     raise UnallowedReturnTypeForUnion
> ```
> 
> means we don't need to iterate over the whole `type_map`, i.e. iterating over` type_.types` is enough. But also I don't understand the internals of Strawberry enough to confidently say there isn't an edge case, maybe once you bring generics into the mix. So I did it was just the safest way possible in terms of breaking changes. If I was a more confident person maybe I would change it and just trust the existing unit tests cover all the relevant cases 😄

Ahaha, same to be honest, we can definitely do this in another PR. I don't remember why I did iterate over all types :'D
